system_metrics_collector/test/topic_statistics_collector/test_received_message_age.cpp

```diff
@@ -0,0 +1,179 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include <chrono>
+#include <string>
+
+#include "topic_statistics_collector/received_message_age.hpp"
+
+#include "rcl/time.h"
+#include "rclcpp/clock.hpp"
+#include "sensor_msgs/msg/imu.hpp"
+#include "std_msgs/msg/string.hpp"
+
+
+namespace
+{
+constexpr const std::chrono::seconds kDefaultDurationSeconds{1};
+constexpr const double kExpectedAverageMilliseconds{2000.0};
+constexpr const double kExpectedMinMilliseconds{1000.0};
+constexpr const double kExpectedMaxMilliseconds{3000.0};
+constexpr const double kExpectedStandardDeviation{816};
+const rclcpp::Time kDefaultSystemTime{0, 0, RCL_SYSTEM_TIME};
+constexpr const int kDefaultTimesToTest{10};
+}  // namespace
+
+class TestReceivedMessageAgeCollector
+  : public topic_statistics_collector::ReceivedMessageAgeCollector<
+    sensor_msgs::msg::Imu>
+{
+public:
+  TestReceivedMessageAgeCollector()
+  {
+    fake_now_ = ReceivedMessageAgeCollector::GetCurrentTime().nanoseconds();
+  }
+  virtual ~TestReceivedMessageAgeCollector() = default;
+
+  /**
+   * Overridden in order to mock the clock for measurement testing.
+   * @return a fixed time point
+   */
+  rclcpp::Time GetCurrentTime() override
+  {
+    return rclcpp::Time{fake_now_, RCL_SYSTEM_TIME};
+  }
+
+  /**
+   * Advance time by a specified duration, in seconds.
+   * @param seconds duration which to advance time
+   */
+  void AdvanceTime(std::chrono::nanoseconds nanos)
+  {
+    fake_now_ += nanos.count();
+  }
+
+  int64_t fake_now_;
+};
+
+sensor_msgs::msg::Imu GetImuMessageWithHeader(const int64_t timestamp)
+{
+  auto message = sensor_msgs::msg::Imu();
+  message.header = std_msgs::msg::Header();
+  message.header.stamp = rclcpp::Time(timestamp);
```

Comment:
```suggestion
  auto message = sensor_msgs::msg::Imu{};
  message.header = std_msgs::msg::Header{};
  message.header.stamp = rclcpp::Time{timestamp};
```
Please ignore this suggestion if these are actually function calls.